### ISSUE TITLE
target selection question/bug

### ISSUE THREAD

**Issue by lspitzner** [haskell/cabal#5273]
With a project setup like:

~~~~
a/a.cabal
a/b/b.cabal
a/c/c.cabal
a/cabal.project
~~~~

where cabal.project has

~~~~
packages: . b c
~~~~

I seem unable to select a specific target and not build the other local packages. For example, if I do this:

~~~~
> cd a/b
> cabal new-build lib:b --constraint='a<0' --constraint='c<0'
~~~~

I run into an error because cabal for some reason makes a, b, and c "user goals"/"targets". This seems to be reproducable on empty/default-init'ialized projects. This is on

~~~~
> cabal --version
cabal-install version 2.2.0.0
compiled using version 2.2.0.1 of the Cabal library 
~~~~

On a related note, I would like to stress (again) that I think it would be very useful if cabal would be more verbose/explicit about what it takes as input, how it interprets that input and why it does the things that it does. In this case I have to turn on `-v3` to get this line "targets: aaa, bbb, ccc" which still does not explain at all where these targets come from.. and on -v3 this is hidden in quite some amount of other dumps.

**Comment by grayjay** [comment]
Yes, the targets on the command line don't affect the targets that are passed to the dependency solver, so the constraints can't be satisfied.  new-build always tries to create an install plan for all targets in order to avoid rerunning the solver or rebuilding packages only because more targets were requested.  new-build only solves for the extra targets, it doesn't build them until they're needed.

**Comment by lspitzner** [comment]
@grayjay thanks for the explanation. I have re-worded the initial description to something a bit more constructive.
